ruma-common/src/push.rs

```diff
@@ -206,6 +211,159 @@ impl Serialize for Action {
     }
 }
 
+/// Optional prefix used by `RoomMemberCountIs`
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum RoomMemberCountPrefix {
+    /// No prefix
+    None,
+    /// Equals
+    Eq,
+    /// Less than
+    Lt,
+    /// Greater than
+    Gt,
+    /// Greater or equal
+    Ge,
+    /// Less or equal
+    Le,
+}
+
+impl Default for RoomMemberCountPrefix {
+    fn default() -> Self {
+        RoomMemberCountPrefix::None
+    }
+}
+
+/// A decimal integer optionally prefixed by one of `==`, `<`, `>`, `>=` or `<=`.
+///
+/// A prefix of `<` matches rooms where the member count is strictly less than the given
+/// number and so forth. If no prefix is present, this parameter defaults to `==`.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct RoomMemberCountIs {
+    prefix: RoomMemberCountPrefix,
+    count: UInt,
+}
+
+impl RoomMemberCountIs {
+    /// A decimal integer with no prefix.
+    pub fn new<T: Into<UInt>>(x: T) -> Self {
+        RoomMemberCountIs { prefix: RoomMemberCountPrefix::None, count: x.into() }
+    }
+
+    /// A decimal integer prefixed by `==`.
+    pub fn eq<T: Into<UInt>>(x: T) -> Self {
+        RoomMemberCountIs { prefix: RoomMemberCountPrefix::Eq, count: x.into() }
+    }
+
+    /// A decimal integer prefixed by `>`.
+    pub fn gt<T: Into<UInt>>(x: T) -> Self {
+        RoomMemberCountIs { prefix: RoomMemberCountPrefix::Gt, count: x.into() }
+    }
+
+    /// A decimal integer prefixed by `>=`.
+    pub fn ge<T: Into<UInt>>(x: T) -> Self {
+        RoomMemberCountIs { prefix: RoomMemberCountPrefix::Ge, count: x.into() }
+    }
+
+    /// A decimal integer prefixed by `<`.
+    pub fn lt<T: Into<UInt>>(x: T) -> Self {
+        RoomMemberCountIs { prefix: RoomMemberCountPrefix::Lt, count: x.into() }
+    }
+
+    /// A decimal integer prefixed by `<=`.
+    pub fn le<T: Into<UInt>>(x: T) -> Self {
+        RoomMemberCountIs { prefix: RoomMemberCountPrefix::Le, count: x.into() }
+    }
+}
+
+impl<T> From<T> for RoomMemberCountIs
+where
+    T: Into<UInt>,
```

Comment:
I don't think there's any precedent for taking `impl Into<ActualType>` parameters in ruma. I'm not a huge fan of this pattern, but would be open to discuss it over in our Matrix room at [#ruma:matrix.org](https://matrix.to/#/#ruma:matrix.org). But for this PR, let's not introduce it yet.

Also, if we're doing construction using `From<UInt>`, I think we should also do `From<RangeFrom<UInt>>`, `From<RangeTo<UInt>>`, `From<RangeInclusive<UInt>>` and `From<RangeToInclusive<UInt>>` in addition to or as a replacement for the above constructors.

Also (last one) if we do add constructors for `RoomMemberCountIs` (which I like), we probably shouldn't make `RoomMemberCountPrefix` public (or is it actually required to be public for some reason?).

Reply:
> `Into`

Yeah, that's just overengineering on my part, sorry 😅 

> `From` traits, constructors and `RoomMemberCountPrefix`

Probably related to https://github.com/ruma/ruma/pull/85#discussion_r443210435.

I intended `RoomMemberCountIs` to be a dumb record with public fields since there are no invariants to guarantee. For that `RoomMemberCountPrefix` does need to be public.

Replacing all the constructors with `From` is a great idea! I have not thought of doing that for all cases, and it took me some time to remember that we can use the range syntax: `::from(2=..)` is even better than `::ge(2)`.

Reply:
Ah, but there's no exclusive `2..` 😕

Reply:
> I intended `RoomMemberCountIs` to be a dumb record with public fields

You didn't make the fields public though ;)

I don't really have a preference as to whether the fields should be private or public. @iinuwa wdyt?

> Ah, but there's no exclusive `2..` 😕

Ah, of course! Let's keep the explicit constructors then. But having the range constructors still seems worthwhile anyway.

Reply:
> You didn't make the fields public though ;)

Yeah, that's an oversight, fixed!

> Ah, of course! Let's keep the explicit constructors then. But having the range constructors still seems worthwhile anyway.

Hmm, I replaced all but two constructors to see if it makes sense to combine these approaches.

I guess it's not super intuitive even with docs because the numbers have to be wrapped in `uint!()`. On the other hand it might depend on context? If all the numbers are already `UInt`, `::from(x..)` might be fine...

Reply:
> Hmm, I replaced all but two constructors

Just `::gt()` now.

Reply:
I haven't written lots of code using our libs myself, but so far nobody has really complained about `UInt` being everywhere in Ruma, even though the constructor macros weren't available until very recently. So I think that should be fine.

I do wonder whether we should have both explicit constructors and `From` implementations too, but am fine merging this as-is.